src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const prevStates = [{ ...state }];
+
+  for (let index = 0; index < actions.length; index++) {
+    if (index > 0) {
+      prevStates[index] = { ...prevStates[index - 1] };
+    }
+
+    switch (actions[index].type) {
+      case 'addProperties':
+        Object.assign(prevStates[index], actions[index].extraData);
+        break;
+
+      case 'removeProperties':
+        actions[index].keysToRemove.forEach(key => {
+          delete prevStates[index][key];
+        });
+        break;
+
+      case 'clear':
+        Object.keys(prevStates[index])
+          .forEach(key => delete prevStates[index][key]);
```

Comment:
Clear means deleting every key of an object, right? Then you could just make the current state reference an empty object, and the previous one would be simply deleted by the garbage collector of a javascript engine)

`prevStates[index] = {}`

Reply:
oh yeah, I was going to but forgot